app/configurator/components/chart-controls/color-ramp.tsx

```diff
@@ -1,34 +1,209 @@
-import { useEffect, useRef } from "react";
-import { interpolateOranges } from "d3";
+import { Trans } from "@lingui/macro";
+import { useSelect } from "downshift";
+import { get } from "lodash";
+import { useEffect, useMemo, useRef } from "react";
+import { Box, Button, Text } from "theme-ui";
+import {
+  DivergingPaletteType,
+  SequentialPaletteType,
+  useConfiguratorState,
+} from "../..";
+import { Label } from "../../../components/form";
+import { Icon } from "../../../icons";
+import { divergingPalettes, sequentialPalettes } from "../ui-helpers";
 // Adapted from https://observablehq.com/@mbostock/color-ramp
-export const ColorRamp = ({
-  colorInterpolator = interpolateOranges,
-  nbClass = 512,
-  width = 288,
-  height = 40,
-}: {
+
+type ColorRampProps = {
   colorInterpolator: (t: number) => string;
-  nbClass: number;
+  nbClass?: number;
   width?: number;
   height?: number;
-}) => {
+};
+
+export const ColorRamp = ({
+  colorInterpolator,
+  nbClass = 512,
+  width = 148,
+  height = 28,
+}: ColorRampProps) => {
   const canvasRef = useRef<HTMLCanvasElement>(null);
-  // Clear by returning a function from usEEFFECT?
+
   useEffect(() => {
     const canvas = canvasRef.current;
     const context = canvas && canvas.getContext("2d");
+
     if (canvas && context) {
-      canvas.style.width = `${width}px`;
-      canvas.style.height = `${height}px`;
+      context.clearRect(0, 0, width, height);
       canvas.style.imageRendering = "-moz-crisp-edges";
       canvas.style.imageRendering = "pixelated";
-      context.clearRect(0, 0, width, height);
-      for (let i = 0; i < nbClass; ++i) {
-        context.fillStyle = colorInterpolator(i / (nbClass - 1));
-        context.fillRect((width / nbClass) * i, 0, width / nbClass, height);
+      canvas.style.borderRadius = "2px";
+
+      const [widthPerClass, numberOfSteps] =
+        nbClass > width ? [1, width] : [width / nbClass, nbClass];
+
+      for (let i = 0; i < numberOfSteps; ++i) {
+        context.fillStyle = colorInterpolator(i / (numberOfSteps - 1));
+        context.fillRect(widthPerClass * i, 0, widthPerClass, height);
       }
     }
+  }, [colorInterpolator, height, width, nbClass]);
+
+  return <canvas ref={canvasRef} width={width} height={height} />;
+};
+
+type ColorRampFieldProps = Omit<ColorRampProps, "colorInterpolator"> & {
+  field: string;
+  path: string;
+};
+
+export const ColorRampField = ({
+  field,
+  path,
+  nbClass,
+}: ColorRampFieldProps) => {
+  const [state, dispatch] = useConfiguratorState();
+
+  const palettes = useMemo(
+    () => [...divergingPalettes, ...sequentialPalettes],
+    []
+  );
+
+  const currentPaletteName = get(
+    state,
+    `chartConfig.fields.${field}.${path}`
+  ) as DivergingPaletteType | SequentialPaletteType;
```

Comment:
Nit: I am not a fan of using get here since we lose type information and have to readd it manually, can't we use optional chaining here ?

Reply:
Hmmm, `field` and `path` are raw strings, so there are some TS conflicts. I could change this so the `field` and `path` would only take pre-defined values (e.g. "areaLayer" and "palette"), but right now we only use this in maps, so I think there also would need to be some kind of check for the chart type inside a `ColorRampField` to make TS happy and then it would not be so generic. I am not sure how to handle this properly - do you have any idea?

Reply:
I think having more narrow types in COlorRampField props and having typeguards in the renderer of ColorRampField so that we make sure that a ColorRampField is not rendered with the wrong types would be good.

Reply:
const ColorRampFieldProps = {
field: MapField,
path: MapPath
}

const Panel = () => {
  return <>
    { isMapField(field) && isMapPath(path) && <ColorFieldRamp field={field} path={path} />  
</>
}
```

something like that ?

Reply:
Yes, but this would limit the ramp picker only to the maps, and maybe it could be also used in other places? But for now I can limit it to maps and we can change this later if needed 👍 Anyway, color types for other charts aren't there yet, so there is still some work to do related to colors :D